kv/dist_sender.go

```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+	replicas := newReplicaSlice(ds.gossip, desc)
+
+	// Rearrange the replicas so that those replicas with long common
+	// prefix of attributes end up first. If there's no prefix, this is a
+	// no-op.
+	order := ds.optimizeReplicaOrder(replicas)
+
+	args := call.Args
+	reply := call.Reply
+
+	// If this request needs to go to a leader and we know who that is, move
+	// it to the front.
+	if !(proto.IsRead(args) && args.Header().ReadConsistency == proto.INCONSISTENT) &&
+		leader.StoreID > 0 {
+		if i := replicas.FindReplica(leader.StoreID); i >= 0 {
+			replicas.MoveToFront(i)
+			order = rpc.OrderStable
+		}
+	}
+
+	rpcErr := ds.sendRPC(desc.RaftID, replicas, order, args, reply)
+
+	// For an RPC error to occur, we've must've been unable to contact
+	// any replicas. In this case, likely all nodes are down (or not getting
+	// back to us within a reasonable amount of time).
+	// We may simply not be trying to talk to the up-to-date replicas, so
+	// clearing the descriptor here should be a good idea.
+	// TODO(tschottdorf): If a replica group goes dead, this will cause clients
+	// to put high read pressure on the first range, so there should be some
+	// rate limiting here.
+	if rpcErr != nil {
+		if log.V(1) {
+			log.Warningf("RPC error; expiring cached descriptor")
+		}
+		ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+	}
+
+	if err := util.FirstError(rpcErr, reply.Header().GoError()); err != nil {
+		if log.V(1) {
+			log.Warningf("failed to invoke %s: %s", call.Method(), err)
+		}
+
+		// If retryable, allow retry. For range not found or range
+		// key mismatch errors, we don't backoff on the retry,
+		// but reset the backoff loop so we can retry immediately.
+		switch tErr := err.(type) {
+		case *proto.RangeNotFoundError, *proto.RangeKeyMismatchError:
+			// Range descriptor might be out of date - evict it.
+			ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+			// On addressing errors, don't backoff; retry immediately.
+			return retry.Reset, nil
+		case *proto.NotLeaderError:
+			newLeader := tErr.GetLeader()
+			// Verify that leader is a known replica; if not,
+			// we've got a stale replica; evict cache.
+			// Next, cache the new leader.
+			if newLeader != nil {
+				if i := replicas.FindReplica(newLeader.StoreID); i == -1 {
+					ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+				}
+			} else {
+				newLeader = &proto.Replica{}
+			}
+			ds.updateLeaderCache(proto.RaftID(desc.RaftID), *newLeader)
+			return retry.Reset, nil
+		default:
+			if retryErr, ok := err.(util.Retryable); ok && retryErr.CanRetry() {
+				return retry.Continue, nil
+			}
+		}
+		return retry.Break, err
+	}
+	return retry.Break, nil
+}
+
+// getDescriptors takes a call and looks up the corresponding range descriptors
+// associated to it. First, the range descriptor for the corresponding StartKey
+// from call.Args is returned; second, if the EndKey exceeds that of the that
```

Comment:
s/the EndKey exceeds that of the that descriptor/the EndKey from call.Args exceeds that of the returned range descriptor/